WIP: Hunspell

mcepl opened the pull request:
It seems to me that hunspell is one of the few successful stories of unification in the free software world. The same spelling engine and same dictionaries can be used now by all Gnome application, all KDE ones, Open/LibreOffice, Firefox/Thunderbird, (optionally) Emacs, and many others.

@caolanm wrote a patch for vim to use hunspell dictionaries directly.

Fixes #846

jamessan commented on src/Makefile:
```diff
@@ -1473,7 +1473,8 @@ ALL_LIBS = \
 	   $(SANITIZER_LIBS) \
-	   $(LEAK_LIBS)
+	   $(LEAK_LIBS) \
+	   $(HUNSPELL_LIBS)
```
You're missing `$(HUNSPELL_CFLAGS)` earlier in the Makefile.  https://github.com/jamessan/vim/commit/09f1bf8d4c721f1b4fe50c22a345d4cf5129e7a3 should cover what needs to be done.

mcepl replied:
Good point. I have not need it, but it is probably The Right Thing™ to do. Added in 387a0a17d (and built as https://copr.fedorainfracloud.org/coprs/mcepl/vim8/build/691979/).

> If Bram is receptive to this work, I can actually dedicate some time to it.

That would be awesome, I gladly admit I have no idea about what I am doing with C programming, so any help is more than welcome.

jamessan commented on src/configure.ac:
```diff
@@ -3080,6 +3080,29 @@ else
+AS_IF([test "x$with_hunspell" != "xno"],
+      [
+       AC_MSG_RESULT($with_hunspell)
+       if pkg-config --exists hunspell ; then
```
I would suggest using pkg-config's autoconf macros instead (e.g., https://github.com/jamessan/vim/commit/4c89bbe0bbaa5772b23c2882c7036728db90fa3e).  That will better handle things like cross-compilation which should be invoking different binaries than the plain `pkg-config`.

mcepl replied:
concerning pkg-config's autoconf macros http://tirania.org/blog/archive/2012/Oct-20.html and besides current ``configure.ac`` doesn't use them (even for Gnome, Gtk, etc. libraries). So, this is expressly no.

Also, using ``PKG_CHECK_MODULES`` didn't work for me, but that isn't actual reason why I stopped using it.

And of course, any suggested fixes are more than welcome. (Hmm, we should probably use ``PKG_CONFIG`` variable, right? Fixed in 3729baa3b).

jamessan replied:
Noted.  Thanks for the information.  #2513 should handle the part I was concerned about.

jamessan commented on src/spell.c:
```diff
@@ -405,6 +405,36 @@ static linenr_T dump_prefixes(slang_T *slang, char_u *word, char_u *pat, int *di
+
+	vim_strncpy(aff + strlen(aff) - 3, "aff", 3);
+
+	lp->sl_hunspell = Hunspell_create(aff, dic);
```
All the calls to the hunspell API are unguarded, so they're being used even when the code wasn't configured to use hunspell.  The commit I linked earlier defines `FEAT_HUNSPELL` which can be used for `#ifdef` checks to disable hunspell-specific code when it isn't needed.

mcepl replied:
Ehm, I don't see ``FEAT_HUNSPELL`` in https://github.com/jamessan/vim/commit/09f1bf8d4c721f1b4fe50c22a345d4cf5129e7a3 , what I do wrong? Any pull requests against my branch https://github.com/mcepl/vim/tree/hunspell are more than welcome, of course.

jamessan replied:
It's in https://github.com/jamessan/vim/commit/4c89bbe0bbaa5772b23c2882c7036728db90fa3e but I hadn't gotten around to the C side of things yet.  I was breaking down the patch into logical commits as I tried to reason through the changes that the patch made.

I didn't finish the branch at the time because I took a detour through implementing [support for enchant](https://github.com/jamessan/vim/tree/pu/enchant-fallback) instead.  I abandoned that eventually because enchant's APIs don't integrate as well with Vim's heuristics and just never circled back to hunspell (which does more closely align with Vim's existing code).

mcepl replied:
I was contemplating enchant as well, but as they say “"We can solve any problem by introducing an extra level of indirection except for the problem of too many levels of indirection.” With Open/LibreOffice we can hope for wide support of ``hunspell`` on many platforms, but ``enchant`` requires supporting yet another library on all those platforms for questionable benefit, so if I am not missing anything, I would rather stay with supporting ``hunspell`` directly.

Perhaps, though, we should cover all this patch with enough ``#ifdef FEAT_HUNSPELL`` and protect the original code, so that we at least preserve backwards compatibility.

jamessan commented on src/spell.c:
```diff
@@ -1902,6 +1975,80 @@ spell_load_lang(char_u *lang)
 
+    if (r == FAIL)
+    {
+#	define HUNSPELLDICT "/usr/share/myspell/"
```
A hard-coded path to the dictionaries is not acceptable.

mcepl replied:
Right, any idea how to discover this location automagically (I wouldn't like to add one more option to ``configure``)? Do we need to parse output of ``echo|hunspell -D``?

robert-scheck replied:
Not sure if I get things right, but could this be made configurable at run-time? And if not configured at run-time, then use the default from hunspell (which maybe could mean `echo|hunspell -D` parsing at build-time, if there is no other/better way).

jamessan reviewed (commented):
Some time ago I started working on porting RedHat's hunspell patch so it was more generally applicable.  Unfortunately I didn't find time to get very far with it, but here are a few comments based on what I had done.

If Bram is receptive to this work, I can actually dedicate some time to it.

mcepl commented:
For the record:
```
gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/atk-1.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib64/dbus-1.0/include -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/libpng15 -I/usr/include/libdrm     -O2 -g -pipe -Wall -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       -I/usr/include/hunspell    -o objects/spell.o spell.c
spell.c: In function 'ensurehunspellinit':
spell.c:415:9: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
         char_u *aff = vim_strnsave(dic, strlen(dic));
         ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:417:2: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
  vim_strncpy(aff + strlen(aff) - 3, "aff", 3);
  ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:417:2: warning: pointer targets in passing argument 2 of 'vim_strncpy' differ in signedness [-Wpointer-sign]
  vim_strncpy(aff + strlen(aff) - 3, "aff", 3);
  ^
In file included from proto.h:135:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/misc2.pro:45:6: note: expected 'char_u *' but argument is of type 'char *'
 void vim_strncpy(char_u *to, char_u *from, size_t len);
      ^
spell.c:419:2: warning: pointer targets in passing argument 1 of 'Hunspell_create' differ in signedness [-Wpointer-sign]
  lp->sl_hunspell = Hunspell_create(aff, dic);
  ^
In file included from spell.h:32:0,
                 from vim.h:2099,
                 from spell.c:75:
/usr/include/hunspell/hunspell.h:12:37: note: expected 'const char *' but argument is of type 'char_u *'
 LIBHUNSPELL_DLL_EXPORTED Hunhandle *Hunspell_create(const char * affpath, const char * dpath);
                                     ^
spell.c:419:2: warning: pointer targets in passing argument 2 of 'Hunspell_create' differ in signedness [-Wpointer-sign]
  lp->sl_hunspell = Hunspell_create(aff, dic);
  ^
In file included from spell.h:32:0,
                 from vim.h:2099,
                 from spell.c:75:
/usr/include/hunspell/hunspell.h:12:37: note: expected 'const char *' but argument is of type 'char_u *'
 LIBHUNSPELL_DLL_EXPORTED Hunhandle *Hunspell_create(const char * affpath, const char * dpath);
                                     ^
spell.c:424:6: warning: pointer targets in passing argument 3 of 'convert_setup' differ in signedness [-Wpointer-sign]
      Hunspell_get_dic_encoding(lp->sl_hunspell)) == FAIL)
      ^
In file included from proto.h:146:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/mbyte.pro:95:5: note: expected 'char_u *' but argument is of type 'char *'
 int convert_setup(vimconv_T *vcp, char_u *from, char_u *to);
     ^
spell.c:430:6: warning: pointer targets in passing argument 2 of 'convert_setup' differ in signedness [-Wpointer-sign]
      Hunspell_get_dic_encoding(lp->sl_hunspell), spell_enc()) == FAIL)
      ^
In file included from proto.h:146:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/mbyte.pro:95:5: note: expected 'char_u *' but argument is of type 'char *'
 int convert_setup(vimconv_T *vcp, char_u *from, char_u *to);
     ^
spell.c: In function 'spell_check':
spell.c:552:6: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
      char_u *mi_final =  mi.mi_word + strlen(mi.mi_word);
      ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:566:4: warning: pointer targets in passing argument 2 of 'Hunspell_spell' differ in signedness [-Wpointer-sign]
    if (Hunspell_spell(lp->sl_hunspell, converted) != 0)
    ^
In file included from spell.h:32:0,
                 from vim.h:2099,
                 from spell.c:75:
/usr/include/hunspell/hunspell.h:22:30: note: expected 'const char *' but argument is of type 'char_u *'
 LIBHUNSPELL_DLL_EXPORTED int Hunspell_spell(Hunhandle *pHunspell, const char *);
                              ^
spell.c:569:8: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
        mi.mi_end = mi.mi_cend = mi.mi_word + strlen(thisword);
        ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:582:3: warning: passing argument 2 of 'spell_iswordp' from incompatible pointer type [enabled by default]
   } while (*mi_end != NUL && spell_iswordp(mi_end, wp->w_buffer));
   ^
spell.c:251:12: note: expected 'struct win_T *' but argument is of type 'struct buf_T *'
 static int spell_iswordp(char_u *p, win_T *wp);
            ^
spell.c: In function 'spell_load_lang':
spell.c:1994:24: warning: pointer targets in initialization differ in signedness [-Wpointer-sign]
   char *resolvedlink = final_name;
                        ^
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
spell.c:1997:3: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:1997:3: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:1997:3: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
spell.c:1997:3: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:1997:3: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:1997:3: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:1997:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:1997:3: warning: pointer targets in passing argument 2 of 'strncmp' differ in signedness [-Wpointer-sign]
   if (strncmp(dp->d_name, lang, strlen(lang)) != 0)
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:143:12: note: expected 'const char *' but argument is of type 'char_u *'
 extern int strncmp (const char *__s1, const char *__s2, size_t __n)
            ^
spell.c:2000:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   if ((strlen(dp->d_name) <= 4) || (dp->d_name[strlen(lang)] != '_'))
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2006:3: warning: pointer targets in passing argument 2 of 'vim_strncpy' differ in signedness [-Wpointer-sign]
   vim_strncpy(spf_name, HUNSPELLDICT, strlen(HUNSPELLDICT));
   ^
In file included from proto.h:135:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/misc2.pro:45:6: note: expected 'char_u *' but argument is of type 'char *'
 void vim_strncpy(char_u *to, char_u *from, size_t len);
      ^
spell.c:2007:3: warning: pointer targets in passing argument 2 of 'vim_strncpy' differ in signedness [-Wpointer-sign]
   vim_strncpy(spf_name + strlen(HUNSPELLDICT), dp->d_name, strlen(HUNSPELLDICT));
   ^
In file included from proto.h:135:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/misc2.pro:45:6: note: expected 'char_u *' but argument is of type 'char *'
 void vim_strncpy(char_u *to, char_u *from, size_t len);
      ^
spell.c:2009:3: warning: pointer targets in passing argument 1 of 'realpath' differ in signedness [-Wpointer-sign]
   if (realpath(spf_name, resolvedlink) != resolvedlink)
   ^
In file included from /usr/include/features.h:375:0,
                 from /usr/include/stdio.h:27,
                 from os_unix.h:21,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/bits/stdlib.h:37:1: note: expected 'const char * __restrict__' but argument is of type 'char_u *'
 __NTH (realpath (const char *__restrict __name, char *__restrict __resolved))
 ^
spell.c:2012:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[0] = tolower(dp->d_name[strlen(lang)+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2012:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[0] = tolower(dp->d_name[strlen(lang)+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2012:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[0] = tolower(dp->d_name[strlen(lang)+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2012:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[0] = tolower(dp->d_name[strlen(lang)+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2012:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[0] = tolower(dp->d_name[strlen(lang)+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2013:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[1] = tolower(dp->d_name[strlen(lang)+1+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2013:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[1] = tolower(dp->d_name[strlen(lang)+1+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2013:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[1] = tolower(dp->d_name[strlen(lang)+1+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2013:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[1] = tolower(dp->d_name[strlen(lang)+1+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2013:3: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
   thisregion[1] = tolower(dp->d_name[strlen(lang)+1+1]);
   ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2019:7: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
       if (thislang[j] && (strcmp(thislang[j]->sl_fname, final_name) == 0))
       ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2019:7: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
       if (thislang[j] && (strcmp(thislang[j]->sl_fname, final_name) == 0))
       ^
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2019:7: warning: pointer targets in passing argument 1 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
       if (thislang[j] && (strcmp(thislang[j]->sl_fname, final_name) == 0))
       ^
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2019:7: warning: pointer targets in passing argument 1 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
       if (thislang[j] && (strcmp(thislang[j]->sl_fname, final_name) == 0))
       ^
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 1 of 'strlen' differ in signedness [-Wpointer-sign]
In file included from os_unix.h:480:0,
                 from vim.h:246,
                 from spell.c:75:
/usr/include/string.h:395:15: note: expected 'const char *' but argument is of type 'char_u *'
 extern size_t strlen (const char *__s)
               ^
spell.c:2019:7: warning: pointer targets in passing argument 1 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
       if (thislang[j] && (strcmp(thislang[j]->sl_fname, final_name) == 0))
       ^
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 1 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2019:7: warning: pointer targets in passing argument 2 of '__builtin_strcmp' differ in signedness [-Wpointer-sign]
spell.c:2019:7: note: expected 'const char *' but argument is of type 'char_u *'
spell.c:2030:7: warning: pointer targets in passing argument 1 of 'vim_strsave' differ in signedness [-Wpointer-sign]
       lp->sl_fname = vim_strsave(resolvedlink);
       ^
In file included from proto.h:135:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/misc2.pro:33:9: note: expected 'char_u *' but argument is of type 'char *'
 char_u *vim_strsave(char_u *string);
         ^
spell.c: In function 'suggest_try_change':
spell.c:4475:3: warning: pointer targets in passing argument 3 of 'Hunspell_suggest' differ in signedness [-Wpointer-sign]
   suggests = Hunspell_suggest(slp->sl_hunspell, &slst, converted);
   ^
In file included from spell.h:32:0,
                 from vim.h:2099,
                 from spell.c:75:
/usr/include/hunspell/hunspell.h:33:30: note: expected 'const char *' but argument is of type 'char_u *'
 LIBHUNSPELL_DLL_EXPORTED int Hunspell_suggest(Hunhandle *pHunspell, char*** slst, const char * word);
                              ^
spell.c:4482:4: warning: pointer targets in passing argument 2 of 'string_convert' differ in signedness [-Wpointer-sign]
    suggest = string_convert(&slp->sl_fromhunconv, slst[i], NULL);
    ^
In file included from proto.h:146:0,
                 from vim.h:2101,
                 from spell.c:75:
proto/mbyte.pro:99:9: note: expected 'char_u *' but argument is of type 'char *'
 char_u *string_convert(vimconv_T *vcp, char_u *ptr, int *lenp);
         ^
```

mcepl committed:
use win_T * as a second parameter of spell_iswordp not buf_T *.

mcepl commented:
Just to add another unresolved problem: when I open new document, spellchecking works perfectly with words from the standard dictionary, but words from user dictionary (``.vim/spell/cs.utf-8.add`` etc.) are marked as wrong. I have to readd one of those words to the user dictionary (which is OK, no duplicate is entered), and then even all words from the user dictionary are recognized.

brammool commented:
It looks like this patch is still far away from a shape where it's ready to include.
Whether it's worth including eventually depends on how well it will work.
When paging through a file, how fast or slow is the spell checking with hunspell?  The main reason for Vim do use its own mechanism was to make this really fast.
Also, how good are suggestions?  The trie mechanism that Vim uses results in far superior suggestions compared to simple mechanisms that try to replace a few characters.

mcepl commented:
@brammool I don't think I have any problems with speed or quality of suggestions (especially with Czech, where vim uses ispell or aspell directly anyway, doesn't it?). I don't edit that long English texts to notice anything and Czech texts are more or less same. A way more dramatic slowdown is from some of my plugins (just now I found that ``ale`` found I have installed ``proselint`` … oh boy!).

However, I don't think the point of this PR is to achieve superior spelling, only to finally achieve the unified spelling experience with all editing on one system corrected by one spelling system (like on some other, ehm, Windows, systems). To that end there are still some problems: for example ``zg`` saves exceptions to the vim's file in ``~/.vim/spell`` not the standard exception file ``~/.hunspell_$LANG``.

I am not a C programmer, so the main purpose of this PR is to attract other people to finish this. The only thing I did was to make autofoo changes necessary so [I can build](https://copr.fedorainfracloud.org/coprs/mcepl/vim8/) with different versions of hunspell.

robert-scheck commented:
@brammool, what is specifically missing that you declare it as far away?

mcepl commented:
@robert-scheck there is a lot unfortunately, see my two points (using incorrect exceptions dictionary, not the standard hunspell's one; see also https://github.com/vim/vim/pull/2500#issuecomment-363580575; whole thing badly needs code review; there are no tests).

brammool commented:
Matej Cepl wrote:

> @brammool I don't think I have any problems with speed or quality of
> suggestions (especially with Czech, where vim uses ispell or aspell
> directly anyway, doesn't it?).

No, Vim has its own spell code, which (at the time I wrote it) is vastly
superior to all existing spell libraries.  aspell comes closest, but
only works well for 8-bit characters, because it uses a hashing
mechanism.

I don't know the current state, since Vim's code has not been improved
since then.  And maintainers haven't updated the files for specific
languages.

It will require someone to actually study the quality of suggestions of
real-life mistakes.  When I worked on it the only data available was for
English, thus that's what it is optimized for.

This is complicated stuff, maybe someone wants to do a masters thesis on
it.

-- 
DEAD PERSON:  I don't want to go in the cart!
CUSTOMER:     Oh, don't be such a baby.
MORTICIAN:    I can't take him...
DEAD PERSON:  I feel fine!
CUSTOMER:     Oh, do us a favor...
MORTICIAN:    I can't.
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

mcepl commented:
> It will require someone to actually study the quality of suggestions of real-life mistakes. When I worked on it the only data available was for English, thus that's what it is optimized for. This is complicated stuff, maybe someone wants to do a masters thesis on it.

Before that master’s thesis is written, could we get to the option ``--with-hunspell`` for ``./configure`` (default off, or ``--with-spell=hunspell`` defaults to ``internal``)? With that you and whoever really cares about the best spellchecking experience (considering the experience is so vastly better, but I really don't want to argue about it) and the rest of us (namely Linux distributions’ maintainers) could have one standard spellchecker? Experience of the rest of us should be same as with LibreOffice, which I would happily accept as sufficient.

And of course, I know, you don't care about what NeoVim does (you shouldn't) and they are not there yet either, but still I see discussion https://github.com/neovim/neovim/issues/356 interesting. ~~Are there any plans to get something like NeoVim’s remote plugins protocol? That could help in communicating with external programs, couldn't it?~~ (That's wrong, hunspell is a library, no external programs are called).

Also, and again I have no idea about internal design, just to tell you that [``:he develop-spell``](https://github.com/b4winckler/vim/blob/master/runtime/doc/develop.txt#L382) was committed Apr 15, 2005. That’s quite a time ago, [hunspell](https://en.wikipedia.org/wiki/Hunspell) hasn’t existed yet (I think, the earliest commit is from Feb 23 09:08:57 2010, but that's import from somewhere else).

And talking about master’s thesis, it seems to me that https://hunspell.github.io/ shows there is actual scholarship work done on hunspell.

mcepl committed:
Replace strlen() with STRLEN() to work with *unsigned char.

mcepl committed:
Add support for hunspell to autoconf.

Commit:
Have vim use hunspell and hunspell dicts directly

It seems to me that hunspell is one of the few successful stories
of unification in the free software world. The same spelling
engine and same dictionaries can be used now by all Gnome
application, all KDE ones, Open/LibreOffice, Firefox/Thunderbird,
(optionally) Emacs, and any others.

Caolan McNamara wrote a patch for vim to use hunspell
dictionaries directly.

His original comment from https://bugzilla.redhat.com/show_bug.cgi?id=219777

    The patch does two things really:

    a) trivially expands the no of regions per language to 20, e.g.
       spanish has a lot of es_XX
    b) makes vim give hunspell dictionaries a go through hunspell for
       spellchecking and suggestions if a native one doesn't exist

Fixes #846

robert-scheck commented:
Honestly, I really would appreciate it if upstream would be less choosey, but help with the here and now to get the current patch included – for the time being (and improve it over the time). It is at least way better than nothing…

brammool commented:
I'm not including a patch that doesn't work.
Also, including a half-finished patch usually means there is less pressure on getting the problems fixed, esp. if the author doesn't care about making it work on different systems.

mcepl commented:
> I'm not including a patch that doesn't work.

Certainly, which is the reason why this MR is marked WIP.

mcepl commented:
OK, apparently nobody is interested enough to help, so giving up, if anybody wants it, ping me and I can reopen, or you can fork my branch https://github.com/mcepl/vim/tree/hunspell directly.

zuxfoucault commented:
a wrapper to aspell https://github.com/shinglyu/vim-codespell

chrisbra commented:
how is that helping for hunspell?

mcepl commented:
> how is that helping for hunspell?

If somebody makes external plugin replacing internal vim spellchecker, it would be trivial to replace aspell with hunspell or anything else (nuspell?).

chrisbra commented:
I am just wondering how is that different to what we did in pre Vim7 times, when all we had was aspell and ispell.
